codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/XmlShapeSerVisitor.java

```diff
@@ -0,0 +1,259 @@
+        TypeScriptWriter writer = context.getWriter();
+        MemberShape memberShape = shape.getMember();
+        Shape target = context.getModel().expectShape(memberShape.getTarget());
+        writer.addImport("XmlNode", "__XmlNode", "@aws-sdk/xml-builder");
+
+        // Use the @xmlName trait if present on the member, otherwise use "member".
+        String locationName = memberShape.getTrait(XmlNameTrait.class)
+                .map(XmlNameTrait::getValue)
+                .orElse("member");
+
+        // Set up a location to store all of the child node(s).
+        writer.write("const collectedNodes: any = [];");
+        // Dispatch to the input value provider for any additional handling.
+        writer.openBlock("(input || []).map(entry => {", "});", () -> {
+            writer.write("const node = $L;", target.accept(getMemberVisitor("entry")));
+            writer.write("collectedNodes.push(node.withName($S));", locationName);
+        });
+
+        writer.write("return collectedNodes;");
+    }
+
+    @Override
+    protected void serializeDocument(GenerationContext context, DocumentShape shape) {
+        throw new CodegenException(String.format(
+                "Cannot serialize Document types on AWS XML protocols, shape: %s.", shape.getId()));
+    }
+
+    @Override
+    protected void serializeMap(GenerationContext context, MapShape shape) {
+        TypeScriptWriter writer = context.getWriter();
+        Model model = context.getModel();
+        writer.addImport("XmlNode", "__XmlNode", "@aws-sdk/xml-builder");
+
+        // Set up a location to store all of the child node(s).
+        writer.write("const collectedNodes: any = [];");
+        // Use the keys as an iteration point to dispatch to the input value providers.
+        writer.openBlock("Object.keys(input).forEach(key => {", "});", () -> {
+            // Prepare a containing node for each entry's k/v pair.
+            writer.write("const entryNode = new __XmlNode(\"entry\");");
+
+            // Prepare the key's node.
+            // Use the @xmlName trait if present on the member, otherwise use "key".
+            MemberShape keyMember = shape.getKey();
+            Shape keyTarget = model.expectShape(keyMember.getTarget());
+            String keyName = keyMember.getTrait(XmlNameTrait.class)
+                    .map(XmlNameTrait::getValue)
+                    .orElse("key");
+            writer.write("const keyNode = new __XmlNode($S');", keyName);
+            writer.write("keyNode.addChildNode($L)", keyTarget.accept(getMemberVisitor("key")));
+            writer.write("entryNode.addChildNode(keyNode);");
+
+            // Prepare the value's node.
+            // Use the @xmlName trait if present on the member, otherwise use "value".
+            MemberShape valueMember = shape.getValue();
+            Shape valueTarget = model.expectShape(valueMember.getTarget());
+            String valueName = valueMember.getTrait(XmlNameTrait.class)
```

Comment:
Followup: there's a lot of logic throughout this CR that repeats the same things over and over. It would be nice to have some kind of XmlIndex that gives you things like computed namespaces and xml names.